jni_bindings/src/lib.rs

```diff
@@ -1,59 +1,55 @@
 #[cfg(target_os = "android")]
 use android_logger::Config;
 use cranelift_entity::EntityRef;
-use jni::objects::{JClass, JIntArray, JObject, JObjectArray, JString, JValue};
-use jni::sys::{jbyte, jint, jlong, jsize};
-use jni::JNIEnv;
-use liveview_native_core::dom::NodeRef;
-use liveview_native_core::ffi::{
-    Attribute, AttributeVec, ChangeType, Element, Node, NodeData, NodeType, RustStr,
+use jni::{
+    objects::{JClass, JIntArray, JObject, JObjectArray, JString, JValue},
+    sys::{jbyte, jint, jlong, jsize},
+    JNIEnv,
+};
+use liveview_native_core::{
+    diff,
+    diff::PatchResult,
+    dom,
+    dom::{Document, NodeRef},
+    ffi::{Attribute, AttributeVec, ChangeType, Element, Node, NodeData, NodeType, RustStr},
 };
-use liveview_native_core::{diff, dom};
 #[cfg(target_os = "android")]
 use log::LevelFilter;
-use std::panic::catch_unwind;
-use std::sync::Mutex;
 
 pub struct JavaResult {
+    /// Raw pointer to Document
     pub val: jlong,
     pub error_msg: String,
 }
 
-#[cfg(target_pointer_width = "32")]
-pub fn jlong_to_pointer<T>(val: jlong) -> *mut T {
-    (val as u32) as *mut T
-}
-
-#[cfg(target_pointer_width = "64")]
-pub fn jlong_to_pointer<T>(val: jlong) -> *mut T {
-    val as *mut T
-}
-
-pub fn from_std_string_jstring<S: AsRef<str>>(x: S, env: JNIEnv) -> JString {
-    env.new_string(x).unwrap()
-}
-
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_drop(
-    _env: JNIEnv,
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_drop(
+    mut env: JNIEnv,
     _: JClass,
+    // non-null raw pointer to Document
```

Comment:
The `non-null` can't be true or there would be no need for the `this.is_null()` check below.

```suggestion
    // raw pointer to Document
```

Reply:
Those changes are as a better means to prevent undefined behavior and to gracefully handle this exception because catch_unwind manages panics and not undefined behavior. If a null pointer is passed, this still will throw an exception, so it's documented this way because what is expected in the safety contract is a valid pointer.